tests/regression/test_mesh_vertexonly.py

```diff
@@ -0,0 +1,333 @@
+    # Galerkin Projection of expression is the same as interpolation of
+    # that expression since both exactly point evaluate the expression.
+    assert np.allclose(f.dat.data_ro, g.dat.data_ro)
+    # Assembly works as expected - global assembly (integration) of a
+    # constant on a vertex only mesh is evaluation of that constant
+    # num_vertices (globally) times
+    f.interpolate(Constant(2))
+    assert np.isclose(assemble(f*dx), 2*num_cells_mpi_global)
+
+
+def vectorfunctionspace_tests(vm, family, degree):
+    # Prep: Get number of cells
+    num_cells_mpi_global = MPI.COMM_WORLD.allreduce(vm.num_cells(), op=MPI.SUM)
+    # Can create function space
+    V = VectorFunctionSpace(vm, family, degree)
+    # Can create functions on function spaces
+    f = Function(V)
+    g = Function(V)
+    # Can interpolate and Galerkin project onto functions
+    x = SpatialCoordinate(vm)
+    f.interpolate(2*as_vector(x))
+    g.project(2*as_vector(x))
+    # Get exact values at coordinates with maintained ordering
+    assert np.shape(f.dat.data_ro)[0] == np.shape(vm.coordinates.dat.data_ro)[0]
+    assert np.allclose(f.dat.data_ro, 2*vm.coordinates.dat.data_ro)
+    # Galerkin Projection of expression is the same as interpolation of
+    # that expression since both exactly point evaluate the expression.
+    assert np.allclose(f.dat.data_ro, g.dat.data_ro)
+    # Assembly works as expected - global assembly (integration) of a
+    # constant on a vertex only mesh is evaluation of that constant
+    # num_vertices (globally) times. Note that we get a vertex cell for
+    # each geometric dimension so we have to sum over geometric
+    # dimension too.
+    gdim = vm.geometric_dimension()
+    if gdim == 1:
+        f.interpolate(Constant((1,)))
+    if gdim == 2:
+        f.interpolate(Constant((1, 1)))
+    if gdim == 3:
+        f.interpolate(Constant((1, 1, 1)))
+    assert np.isclose(assemble(inner(f, f)*dx), num_cells_mpi_global*gdim)
+
+
+"""Families and degrees to test function spaces on VertexOnlyMesh"""
+families_and_degrees = [
+    ("DG", 0),
+    pytest.param("CG", 1, marks=pytest.mark.xfail(reason="unsupported family and degree"))
+]
+
+
+@pytest.mark.parametrize("parentmesh", parentmeshes)
+@pytest.mark.parametrize("n", ncoords)
+@pytest.mark.parametrize(("family", "degree"), families_and_degrees)
+def test_functionspaces(parentmesh, n, family, degree):
+    vertexcoords = random_coords(n, parentmesh.geometric_dimension())
+    vm = VertexOnlyMesh(parentmesh, vertexcoords)
+    functionspace_tests(vm, family, degree)
+    vectorfunctionspace_tests(vm, family, degree)
+
+
+@pytest.mark.parallel
+@pytest.mark.parametrize("parentmesh", parentmeshes)
+@pytest.mark.parametrize("n", ncoords)
+@pytest.mark.parametrize(("family", "degree"), families_and_degrees)
+def test_functionspaces_parallel(parentmesh, n, family, degree):
+    test_functionspaces(parentmesh, n, family, degree)
```

Comment:
Prefer introducing a new directory `tests/vertexonly` and splitting some of these tests into different files.

Reply:
Done as of 41ff5a26099e23b42ba7b63a2f38e0876415e37a